### PR TITLE
For dad to do a CR

### PR DESCRIPTION
**Pull request by Snusifer**

### CODE REVIEW

File: src/BlaaResept.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+public final class BlaaResept extends Resept {
+    private final Farge farge = Farge.BLAA;
+    private final double RABATT_DESIMAL = 0.75;
+
+    public BlaaResept(Legemiddel legemiddel, Lege utskrivendeLege, Pasient p, int reit) {
```
Comment (gordost) [inline src/BlaaResept.java]: final Legemiddel legemiddel etc...

File: src/BlaaResept.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        super(legemiddel, utskrivendeLege, p, reit);
+        this.type = Type.BLAA;
+    }
+    @Override
```
Comment (gordost) [inline src/BlaaResept.java]: Consider separation of methods with blank lines, for improved readibility

File: src/Database.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import java.util.Map;
+import java.util.TreeMap;
+
+public class Database {
+	SortertLenkeliste<Legemiddel> legemiddelList = new SortertLenkeliste<>();
```
Comment (gordost) [inline src/Database.java]: Add *private final* to all the variables

File: src/Database.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	public SortertLenkeliste<Resept> getReseptList() {return reseptList;}
+	public SortertLenkeliste<Lege> getLegeList() {return legeList;}
+	
+	public Map<Integer, Legemiddel> getLegemiddelMap() {return legemiddelMap;}
+	public Map<Integer, Pasient> getPasientMap() {return pasientMap;}
+	public Map<Integer, Resept> getReseptMap() {return reseptMap;}
+	public Map<Integer, Lege> getLegeMap() {return legeMap;}
+
+	
+	public void addLegemiddel(Legemiddel legemiddel) {
```
Comment (gordost) [inline src/Database.java]: final Legemiddel legemiddel multiple places below

File: src/Database.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	}
+	public void addResept(Resept resept) {
+		reseptMap.put(resept.hentId(), resept);
+		reseptList.leggTil(resept);
+	}
+	public void addLege(Lege lege) {
+		legeMap.put(lege.hentId(), lege);
+		legeList.leggTil(lege);
+	}
+	/*
```
Comment (gordost) [inline src/Database.java]: Remove the block comment (it's ugly)

File: src/Fastlege.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+final class Fastlege extends Lege implements Kommuneavtale {
+    private final int avtalenummer;
+    
+    public Fastlege(String navn, int avtalenummer) {
```
Comment (gordost) [inline src/Fastlege.java]: final

File: src/Fastlege.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+        super(navn);
+        this.type = "Fastlege";
```
Comment (gordost) [inline src/Fastlege.java]: Consider using enums whenever possible

File: src/HvitResept.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+public class HvitResept extends Resept {
+    private Farge farge = Farge.HVIT;
```
Comment (gordost) [inline src/HvitResept.java]: final

File: src/HvitResept.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+    public HvitResept(Legemiddel legemiddel, Lege utskrivendeLege, Pasient p, int reit) {
```
Comment (gordost) [inline src/HvitResept.java]: final arguments

File: src/HvitResept.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+        super(legemiddel, utskrivendeLege, p, reit);
+        this.type = Type.HVIT;
+    }
+    @Override
```
Comment (gordost) [inline src/HvitResept.java]: Add blank lines between the method bodies below

File: src/Lege.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
+class Lege implements Comparable<Lege> {
+    private static int statiskId = 0;
+    private final int id;
+    private final String navn;
+    private Lenkeliste<Resept> reseptList = new Lenkeliste<>();
+    protected String type;
```
Comment (gordost) [inline src/Lege.java]: enum

File: src/Lege.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+    public Lege(String navn) {
```
Comment (gordost) [inline src/Lege.java]: final

File: src/Lege.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
+class Lege implements Comparable<Lege> {
+    private static int statiskId = 0;
+    private final int id;
+    private final String navn;
+    private Lenkeliste<Resept> reseptList = new Lenkeliste<>();
```
Comment (gordost) [inline src/Lege.java]: final

File: src/Lege.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    protected String type;
+
+    public Lege(String navn) {
+    		this.id = getNesteId();
+        this.navn = navn;
+        this.type = "Lege";
+    }
+
+    public int hentId() {return id;}
```
Comment (gordost) [inline src/Lege.java]: Blank lines between methods

File: src/Legemiddel.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+			for (Type t : Type.values()) {
+				if (t.name.equals(name)) {
+					return t;
+				}
+			}
+			throw new RuntimeException("Not found.");
+		}
+		
+    };
+    protected Type type;
```
Comment (gordost) [inline src/Legemiddel.java]: final

File: src/Legemiddel.java [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+
+    public Legemiddel(String navn, double pris, double virkestoff) {
```
Comment (gordost) [inline src/Legemiddel.java]: final parameters

File: src/Legemiddel.java [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    }
+    public Type hentType() 			    		{return type;}
+    public int hentId()                     {return id;}
+    public String hentNavn()                {return navn;}
+    public double hentPris()                {return pris;}
+    public double hentVirkestoff()          {return virkestoff;}
+    public abstract int hentStyrke();
+    public abstract int hentCompareInt();
+    public void settNyPris(double nyPris)   {pris = nyPris;}
+    @Override
```
Comment (gordost) [inline src/Legemiddel.java]: Blank lines between methods (doesn't apply to one-liner methods, they are OK)

File: src/Legemiddel.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    public int compareTo(Legemiddel annetLegemiddel) {
```
Comment (gordost) [inline src/Legemiddel.java]: final argument

File: src/LegemiddelB.java [hunk 19]
```diff
@@ -0,0 +1,18 @@
+public final class LegemiddelB extends Legemiddel {
+    private final int vanedannendeStyrke;
+    private final int compareInt = 0;
+
+    public LegemiddelB(String navn, double pris, double virkestoff, int vanedannendeStyrke) {
```
Comment (gordost) [inline src/LegemiddelB.java]: final arguments, line breaks if the list gets too long

File: src/LegemiddelC.java [hunk 20]
```diff
@@ -0,0 +1,16 @@
+public final class LegemiddelC extends Legemiddel {
+	private final int compareInt = 1;
+
+
+    public LegemiddelC(String navn, double pris, double virkestoff) {
```
Comment (gordost) [inline src/LegemiddelC.java]: final arguments

**Comment by Snusifer** [comment]
Alle er handlet

File: src/Controller.java [hunk 21]
```diff
@@ -160,9 +156,9 @@ private void leggTilLegemiddel(Legemiddel.Type type) {
 		default:
 			throw new RuntimeException("Ugyldig type " + type);
 		}
-		System.out.println("\nTast inn b for å gå tilbake.");
+		System.out.println("\nTast inn Enter for å gå tilbake.");
 		String ui = inputScanner.next();
-		if (ui.equals("b")) {
+		if (ui.equals("")) {
```
Comment (gordost) [inline src/Controller.java]: It is better to write `if ("".equals(ui))`

File: src/Controller.java [hunk 22]
```diff
@@ -255,9 +251,9 @@ private void leggTilLege() {
 	private void leggTilPasient() {
 		printEmptyLines();
 		System.out.println("# leggTilpasient (navn, fnr)");
-		System.out.println("\nTast inn b for å gå tilbake.");
+		System.out.println("\nTast inn Enter for å gå tilbake.");
 		String ui = inputScanner.next();
-		if (ui.equals("b")) {
+		if (ui.equals("")) {
```
Comment (gordost) [inline src/Controller.java]: same here

File: src/Controller.java [hunk 23]
```diff
@@ -305,9 +301,9 @@ private void leggTilResept(Resept.Type type) {
 		System.out.println("Tast inn l for lege-oversikt.");
 		System.out.println("Tast inn p for pasient-oversikt.");
 		System.out.println("Tast inn a for alle-oversikt.");
-		System.out.println("\nTast inn b for å gå tilbake.");
+		System.out.println("\nTast inn Enter for å gå tilbake.");
 		String ui = inputScanner.next();
-		if (ui.equals("b")) {
+		if (ui.equals("")) {
```
Comment (gordost) [inline src/Controller.java]: And here

File: src/Controller.java [hunk 24]
```diff
@@ -479,11 +475,13 @@ private void velgReseptOgBrukReit (Pasient chosenPasient) {
 				chosenResept = r;
 				printEmptyLines();
 				if(chosenResept.bruk()) {
-					System.out.println("Bruker reit av resept id: " + chosenResept.hentId() + ". (" + chosenResept.hentLegemiddel().hentNavn() + "). Gjenstående reit er: " + chosenResept.hentReit());
+					System.out.println("Bruker reit av resept id: " + chosenResept.hentId() + ". (" + chosenResept.hentLegemiddel().hentNavn() + "). Pasient: (" + chosenPasient.hentNavn() + "). Gjenstående reit er: " + chosenResept.hentReit());
```
Comment (gordost) [inline src/Controller.java]: Consider breaking this line, it's too long

File: src/Controller.java [hunk 25]
```diff
@@ -479,11 +475,13 @@ private void velgReseptOgBrukReit (Pasient chosenPasient) {
...
 				} else {
-					System.out.println("Forsøkte å bruke reit av resept id: " + chosenResept.hentId() + ". (" + chosenResept.hentLegemiddel().hentNavn() + "). Resepten er ugyldig.");
+					System.out.println("Forsøkte å bruke reit av resept id: " + chosenResept.hentId() + ". (" + chosenResept.hentLegemiddel().hentNavn() + "). Pasient: (" + chosenPasient.hentNavn() + "). Resepten er ugyldig.");
```
Comment (gordost) [inline src/Controller.java]: Same here

File: src/Controller.java [hunk 26]
```diff
@@ -479,11 +475,13 @@ private void velgReseptOgBrukReit (Pasient chosenPasient) {
...
 				}
-				inputScanner.next();
+				System.out.println("\nTast Enter for å gå tilbake. Tast r for å bruke en til.");
+				ui = inputScanner.next();
+				if (ui.equals("r")) {velgPasient();}
```
Comment (gordost) [inline src/Controller.java]: Better `if ("r").equals(ui)`

File: src/Controller.java [hunk 27]
```diff
@@ -575,20 +573,24 @@ else if (parts[1].charAt(0) == 'c') {
...
+					case MILITAER:
+						reit = Integer.parseInt(parts[4]);
+						database.addResept(new MilitaerResept(legemiddel, lege, pasient, reit));
+						break;
+					case P:
+						database.addResept(new PResept(legemiddel, lege, pasient));
+						break;		
 				}
 			}
 		} catch (NumberFormatException|ArrayIndexOutOfBoundsException|NoSuchElementException e) {
```
Comment (gordost) [inline src/Controller.java]: Leave spaces arount `|`

File: src/Controller.java [hunk 28]
```diff
@@ -664,6 +669,9 @@ private void stat2() {
 		}
 		System.out.println("Totalt antall utskrevne resepter på vanedannende legemidler til militæret: " + list.stoerrelse() + "\n");
```
Comment (gordost) [inline src/Controller.java]: Line too long - break it!

File: src/Controller.java [hunk 29]
```diff
@@ -676,6 +684,9 @@ private void stat3() {
 				System.out.println(l.hentId() + " - " + l.hentNavn() + ", " + l.hentAvtalenummer() + ", Antall narkotiske resepter: " + antall);
```
Comment (gordost) [inline src/Controller.java]: Same here

File: src/Lege.java [hunk 30]
```diff
@@ -23,26 +22,33 @@ public int hentAntNarkResept() {
     		}
     		return c;
     }
+    
     @Override
     public int compareTo(Lege annenLege) {
     		return navn.compareTo(annenLege.hentNavn());
     }
+    
     public int compareTo(int annenLegeId) {
```
Comment (gordost) [inline src/Lege.java]: final

File: src/Lege.java [hunk 31]
```diff
@@ -23,26 +22,33 @@ public int hentAntNarkResept() {
...
     }
+    
     public Lenkeliste<Resept> getReseptList() {return reseptList;}
+    
     public Resept addResept(Resept resept) {
     		reseptList.leggTil(resept);
     		return resept;
     }
+    
     public Resept addResept(int pos, Resept resept) {
```
Comment (gordost) [inline src/Lege.java]: final

File: src/Legemiddel.java [hunk 32]
```diff
@@ -31,28 +31,36 @@ public static Type of(final String name) {
...
+    
     public String hentNavn()                {return navn;}
+    
     public double hentPris()                {return pris;}
+    
     public double hentVirkestoff()          {return virkestoff;}
-    public abstract int hentStyrke();
-    public abstract int hentCompareInt();
+    
     public void settNyPris(double nyPris)   {pris = nyPris;}
```
Comment (gordost) [inline src/Legemiddel.java]: final

**Pull request merged**